protocols/kad/src/behaviour.rs

```diff
@@ -2133,40 +2228,39 @@ where
                 user_data,
             } => {
                 if let Some(query) = self.queries.get_mut(&user_data) {
+                    let stats = query.stats().clone();
                     if let QueryInfo::GetRecord {
                         key,
-                        records,
-                        quorum,
+                        ref mut count,
                         cache_candidates,
+                        ref mut record_to_cache,
                     } = &mut query.inner.info
                     {
                         if let Some(record) = record {
-                            records.push(PeerRecord {
+                            if record_to_cache.is_none() {
+                                *record_to_cache = Some(record.clone());
+                            }
```

Comment:
Say we discover multiple records for the same key. With the above only the first record is replicated. `libp2p-kad` can not know which of the records is the one to be cached, i.e. it can not make a qualified judgement on which of the records is the best one.

How about removing this automated publishing mechanism and depend on the user to call `Kademlia::put_record_to` instead?

https://github.com/libp2p/rust-libp2p/blob/6da8497af5c3ce076534be68116c01b7f1748383/protocols/kad/src/behaviour.rs#L778-L798

Either way, the comment on `put_record_to` needs updating.

Reply:
I have removed the caching, but I am confused why the behaviour of `put_record_to` changed if I only changed the behaviour of `get_record`, does that rely on that caching to do this?

Reply:
According to the Kademlia paper:

> For caching purposes, once a lookup succeeds, the requesting node stores the (key, value) pair at the closest node it observed to the key that did not return the value.

Before this pull request, when we did a `Kademlia::get_record` with a quorum of 1, we could replicate the (key, value) pair to the closest node that did not have the pair. But if we did a quorum of >1, the `Kademlia::get_record` query would potentially discover more than one value for the given key. Which value would you replicate to the closest node that did not have the value? `libp2p-kad` can not make that choice. With a quorum of >1 `Kademlia::get_record` would thus not replicate the pair to the closest node that did not have the value. Instead it relied on the user first picking the right value and then calling `Kademlia::put_record_to`.

This pull request removes the quorum mechanism. To keep things simple, I suggest also removing the automated replication to the closest node that does not have the value, but instead entirely rely on the user to call `put_record_to`.

Does that make sense @dignifiedquire? Happy to expand further.

Reply:
I believe https://github.com/libp2p/rust-libp2p/pull/2712/commits/d8ac894f573955c9eef914f7b832798f16e6eb3d does what you requested